src/libutil/args.hh

```diff
@@ -110,7 +120,30 @@ protected:
         { }
     };
 
-    /* Options. */
+    /**
+     * The basic function type of the completion callback.
+     *
+     * Used to define CompleterClosure and some common case completers
+     * that individual flags/arguments can use.
+     *
+     * The AddCompletions that is passed is an interface to the state stored as part of the root command
+     */
+    typedef void CompleterFun(AddCompletions &, size_t, std::string_view);
+
+    /**
+     * The closure type of the completion callback.
+     *
+     * This is what is actually stored as part of each Flag / Expected
+     * Arg.
+     */
+    typedef std::function<CompleterFun> CompleterClosure;
```

Comment:
Storing a function means it can't close over anything. So we usually *store* the closure, but *use* the plain function type (which cause be a regular top-level function or closure).

This is like the Vector vs Slice distinction in rust (or `std::vector` vs arrays in c++).